chain-signatures/node/src/indexer_common.rs

```diff
@@ -0,0 +1,455 @@
+            SignatureRespondedEvent::Hydration(event) => event.request_id,
+        }
+    }
+
+    pub fn signature(&self) -> Signature {
+        match self {
+            SignatureRespondedEvent::Solana(event) => {
+                crate::indexer_sol::to_mpc_signature(event.signature.clone()).unwrap()
+            }
+            SignatureRespondedEvent::Hydration(event) => event.signature.clone(),
+        }
+    }
+}
+
+pub(crate) trait SignatureEventTrait {
+    fn generate_request_id(&self) -> [u8; 32];
+    fn generate_sign_request(
+        &self,
+        entropy: [u8; 32],
+        total_timeout: Duration,
+    ) -> anyhow::Result<IndexedSignRequest>;
+    fn source_chain(&self) -> Chain;
+    fn sender_string(&self) -> String;
+}
+
+pub(crate) trait SignatureEvent: SignatureEventTrait + std::fmt::Debug {}
+
+pub(crate) type SignatureEventBox = Box<dyn SignatureEvent + Send>;
+
+pub(crate) async fn process_sign_event(
+    sign_event: SignatureEventBox,
+    entropy: [u8; 32],
+    sign_tx: mpsc::Sender<Sign>,
+    node_near_account_id: AccountId,
+    total_timeout: Duration,
+    backlog: Backlog,
+) -> anyhow::Result<()> {
+    let sign_request = sign_event.generate_sign_request(entropy, total_timeout)?;
+
+    // Insert the transaction into the backlog when we first see the sign request
+    let sign_id = sign_request.id;
+    let sign_request_type = sign_request.sign_request_type.clone();
+
+    // Create the appropriate BacklogTransaction based on the sign request type
+    let backlog_tx = match &sign_request_type {
+        SignRequestType::Sign => BacklogTransaction::Sign(SignTx {
+            request_id: sign_id.request_id,
+            source_chain: sign_event.source_chain(),
+            key_version: sign_request.args.key_version,
+            status: PendingRequestStatus::AwaitingResponse,
+        }),
+        SignRequestType::SignBidirectional(_event) => {
+            // For bidirectional requests, start with a Sign transaction
+            // The protocol will advance it to Bidirectional after generating the signature
+            BacklogTransaction::Sign(SignTx {
+                request_id: sign_id.request_id,
+                source_chain: sign_event.source_chain(),
+                key_version: sign_request.args.key_version,
+                status: PendingRequestStatus::AwaitingResponse,
+            })
+        }
+        _ => anyhow::bail!("Unexpected sign request type"),
+    };
+
+    backlog
+        .insert(
+            sign_event.source_chain(),
+            sign_id,
+            backlog_tx,
+            sign_request_type,
+        )
+        .await;
+
+    if let Err(err) = sign_tx.send(Sign::Request(sign_request)).await {
+        // TODO: handle error to ensure 100% success rate
+        let chain = sign_event.source_chain();
+        tracing::error!(?err, chain = %chain, "Failed to send {} sign request into queue", chain.as_str());
+    } else {
```

Comment:
Hm, so if this fails, we have a full channel buffer. A burst of incoming requests could easily trigger this. So anyone could try and abuser this to prevent a signature from getting through.

But the good news is, at this point we already have the request in the backlog. All we need is to ensure we start new attempts on backlog requests if no success is seen for a time. (cc @ChaoticTempest )

@ppca  Can you link this to issue or something to make it less likely this is just buried and forgotten?